keps/sig-storage/2848-transfer-volumesnapshots-namespaces/README.md

```diff
@@ -0,0 +1,795 @@
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+This KEP proposes a method for transferring VolumeSnapshot between namespaces. 
+
+Currently, VolumeSnapshot, and VolumeSnapshotContent are bound to specific namespace. The binding relationship between these two storage resources is one-to-one.
+This proposal targets allowing users to conveniently move these resources without manual creation
+of each independent object (VolumeSnapshot, VolumeSnapshotContent).
+
+Based on community design sessions, and inspired by [API for PVC Namespace Transfer](https://github.com/kubernetes/enhancements/pull/2110/files), the following is being proposed :-
+
+- `VolumeSnapshotTransferRequest` resources are created in the target namespace.
+- `VolumeSnapshotTransferApproval` resources are created in the source namespace.
+-  When a controller detects matching request/approval resources, a transfer in initiated.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+Allow Kubernetes users to quickly and conveniently transfer VolumeSnapshots across namespaces.
+### Goals
+- Use CRDs for VolumeSnapshot Namespace Transfer request/approval.
+- Edit external SnapshotController to accept/reject/init VolumeSnapshot transfer.
+- The source and target VolumeSnapshot will point to the same VolumeSnapshotContent.
+- Use Validation Webhook to reject any transfer contains secret.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Transfer of resources other than VolumeSnapshot will not be discussed.
+- Transfer of VolumeSnapshots which contain secret.
```

Comment:
I think that we should discuss this. Transfer snapshot without secret can be a milestone for alpha, for example. However, I guess that snapshot with secret will need to be supported before GA. 

Also, I think that secrets defined without template like `$namespace` should work well even after transfer in most cases. So, it can be added as a scope for initial implementation.

Reply:
+1, I think we should have solid design of secrets handling. It does not need to be in the initial implementation, but we don't want to realize the approach we design here is wrong.

Reply:
To make `StorageTransferRequest` to be a generic interface and to handle resource specific requirement, we will need to add something like `spec.options` field to allow key/value pairs (map[string]string) to be passed by users. By using this field, users can pass resource specific options, like whether to overwrite secrets, through `StorageTransferRequest` to the transfer controller of each resource.

```
spec:
  options:
    secretOverwrite : "false"
```

And by doing so, we might be able to decide not implementing `secretOverwrite = true` case for alpha, with a room to provide this feature later without changing the CRD definition.

Also, the options can be resource specific, like below:
```
spec:
  options:
    snapshotterSecretName : "newSecret"
    snapshotterSecretNamespace : "destinationNamespace"
```

Reply:
> To make `StorageTransferRequest` to be a generic interface and to handle resource specific requirement, we will need to add something like `spec.options` field to allow key/value pairs (map[string]string) to be passed by users. By using this field, users can pass resource specific options, like whether to overwrite secrets, through `StorageTransferRequest` to the transfer controller of each resource.

@mkimuram @jsafrane in the design session we agreed to have a validation webhook to reject the transfer in case it contains secrets

Reply:
Let me explain more details on my intention of my comments [here](https://github.com/kubernetes/enhancements/pull/2849#issuecomment-897759728) and [here]( https://github.com/kubernetes/enhancements/pull/2849#discussion_r681989203).

If snapshotter-secret is defined with a fixed value like below in volumeSnapshotClass:
```
csi.storage.k8s.io/snapshotter-secret-name: shared-secret
csi.storage.k8s.io/snapshotter-secret- namespace: default
```

Admin's intention will be that everyone in the cluster is allowed to use the same secret, or shared-secret in default namespace in above example, because volumeSnapshotClass is cluster-scoped, as a result, everyone in the cluster who used the class can/will take a snapshot by using the same secret. Therefore, no one should care if the secret makes issue after transfer.

On the other hand, if snapshotter-secret is defined with a template like below in volumeSnapshotClass:
- (1) per namespace
```
csi.storage.k8s.io/snapshotter-secret-name: per-namespace-secret
csi.storage.k8s.io/snapshotter-secret- namespace: ${volumesnapshot.namespace}
```
or 
- (2) per snapshot
```
csi.storage.k8s.io/snapshotter-secret-name: ${volumesnapshot.name}
csi.storage.k8s.io/snapshotter-secret- namespace: ${volumesnapshot.namespace}
```

Admin's intention will be to use a different secret per namespace or per snapshot, because the secret used to take a snapshot will be different per namespace or per snapshot. In this situation, whether to allow using the secret after transfer depends on each transfer. For example, when Alice in namespace ns-a transfers her snapshot-a to Bob in namespace ns-b. She may allow Bob to access her per-namespace-secret or snapshot-a in ns-a, or may not allow it.
(Note that Bob won't be able to directly see Alice's secret, just the driver accesses to it when handling the snapshot).

With above in mind, we can say:
- (a) No secret case: No need to care about secrets
- (b) Secret with fixed value case: No need to care about secrets after transfer make issue
- (c) Secret with templated value case ((1) and  (2) above): It depends. Keeping secret as is maybe OK, if the user agrees. But if not, it needs to be changed somehow.

Therefore, I'm suggesting not to only focus on (a), but to include (b) as a scope.
And if possible, I'm not saying that it is mandatory, we should consider (c) on how it could be handled later. At least, we can define an option to specify whether to allow keeping secret as is or not, and transferring snapshot with keeping secret as is (And we can make default to be allow and leave disallow case not implemented for alpha).

P.S.
After I wrote above, I started to think whether updating secret after transfer, by using the template makes sense in case (1).
There should be another secret with the same name and different contents in the destination namespace. And even if not, sharing the contents of the secret in the source namespace to the destination namespace won't be a good idea, because it should be a "secret" for the particular namespace (There will be more security risk by sharing the contents than sharing the reference). Although, this idea should be still valid for case (2).    

Maybe, if users really care about sharing secrets, they will need to take a snapshot with snapshot class with per snapshot secret, and transfer the snapshot after sharing the secret in some ways (And this idea would be hard to apply to PVC, but it will be out of scope for now).